### PR TITLE
[OF-1195] feat!: Add XErrorCodes to WebService

### PR DESCRIPTION
**Pull request by Mag-JB**
connected-spaces-platform Pull Request

**For the reviewer**

* [ ] If required, are the changes covered by appropriate tests?
* [ ] Are any public-facing API changes well documented?
* [ ] Is the code easily readable and extensible and/or follow existing conventions?

### CODE REVIEW

File: Library/src/Services/WebService.cpp [hunk 1]
```diff
@@ -104,9 +105,16 @@ void ResultBase::SetResult(csp::services::EResultCode ResCode, uint16_t HttpResC
...
+EResultBaseFailureReason ResultBase::ParseErrorCode(const csp::common::String& Value)
 {
-	return (int) EResultBaseFailureReason::Unknown;
+	if (XErrorCodeToFailureReason.HasKey(Value))
+	{
+		return XErrorCodeToFailureReason[Value];
+	}
+	else
+	{
+		return EResultBaseFailureReason::Unknown;
```
Comment (MAG-SamBirley) [inline Library/src/Services/WebService.cpp:116]: Worth including a log message here to help us catch error codes that we don't yet track in CSP?

**Review by MAG-SamBirley — CHANGES_REQUESTED** [review]

File: Library/include/CSP/Services/WebService.h [hunk 2]
```diff
@@ -105,14 +139,14 @@ class CSP_API ResultBase
...
-	int GetFailureReason() const;
+	EResultBaseFailureReason GetFailureReason() const;
 
 protected:
 	ResultBase(csp::services::EResultCode ResCode, uint16_t HttpResCode);
 
 	void SetResult(csp::services::EResultCode ResCode, uint16_t HttpResCode);
 
-	virtual int ParseErrorCode(const csp::common::String& Value);
+	virtual EResultBaseFailureReason ParseErrorCode(const csp::common::String& Value);
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Services/WebService.h:149]: Does this still need to be virtual?

File: Library/src/Services/WebService.cpp (hunk 1 above)
Comment (MAG-SamBirley) [reply Library/src/Services/WebService.cpp]: Saw the new error message. Super useful! Thanks :)

File: Library/include/CSP/Services/WebService.h (hunk 2 above)
Comment (MAG-MichaelK) [reply Library/include/CSP/Services/WebService.h]: No, and it should also be made private.

File: Library/include/CSP/Services/WebService.h [hunk 3]
```diff
@@ -121,7 +155,44 @@ class CSP_API ResultBase
 	float ResponseProgress = 0.0f;
 
 	csp::common::String ResponseBody;
-	int FailureReason;
+	EResultBaseFailureReason FailureReason;
+
+private:
+	csp::common::Map<csp::common::String, EResultBaseFailureReason> XErrorCodeToFailureReason = {
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Services/WebService.h:161]: This map should be moved to WebService.cpp in an anonymous namespace, changes to a `std::map`, and be made static. We don't want to re-create this every time a new ResultBase instance is made. We should also only be mapping the quota errors and any error codes we already handle as that is all we need to cover in this ticket.

File: Library/include/CSP/Services/WebService.h [hunk 4]
```diff
@@ -62,7 +63,40 @@ enum class EResultCode : uint8_t
 enum class EResultBaseFailureReason
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Services/WebService.h:63]: Can we get a better name for this now that it contains all error codes?

**Review by MAG-MichaelK — CHANGES_REQUESTED** [review]

File: Library/include/CSP/Services/WebService.h (hunk 4 above)
Comment (MAG-MichaelK) [reply Library/include/CSP/Services/WebService.h:63]: Something like `EFailureReason`

**Review by MAG-SamBirley — APPROVED** [review]
Happy with the changes made to address my requests, so approved from that point of view. I would like to also see the feedback from @MAG-MichaelK addressed too though!

**Review by MAG-MichaelK — APPROVED** [review]

**Comment by Mag-JB** [comment]
I will test this before merging!

File: Library/include/CSP/Services/WebService.h [hunk 5]
```diff
@@ -105,14 +138,14 @@ class CSP_API ResultBase
...
-	int GetFailureReason() const;
+	ERequestFailureReason GetFailureReason() const;
 
 protected:
 	ResultBase(csp::services::EResultCode ResCode, uint16_t HttpResCode);
 
 	void SetResult(csp::services::EResultCode ResCode, uint16_t HttpResCode);
 
-	virtual int ParseErrorCode(const csp::common::String& Value);
+	ERequestFailureReason ParseErrorCode(const csp::common::String& Value);
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Services/WebService.h:148]: Can we also make this private as it doesn't need to be accessed by any derived types?

**Review by MAG-MichaelK — CHANGES_REQUESTED** [review]

**Review by MAG-SamBirley — DISMISSED** [review]

File: Library/include/CSP/Services/WebService.h (hunk 5 above)
Comment (MAG-MichaelK) [reply Library/include/CSP/Services/WebService.h:148]: Just waiting on this to be done then I can approve this PR

Comment (Mag-JB) [reply Library/include/CSP/Services/WebService.h:148]: oops didn't see this one sneak in after the approval

**Review by MAG-MichaelK — APPROVED** [review]

**Review by MAG-AdamThorn — APPROVED** [review]

**Pull request merged**